tests/unit/memefficiency.tcl

```diff
@@ -397,13 +405,141 @@ run_solo {defrag} {
+            set channel_name_dict [dict create] ;# dict to store the various channel name of different sizes,
+                                                ;# the keys correspond to the sizes in channel_sizes
+
+            # Create channels of increasing size, starting from 400 and increasing by
+            # 50 till we reach the maximum size. The purpose of this is to allow us to
+            # use a single client to fill in gaps between different keys, which makes
+            # fragmentation more likely. If we don't do this, we would need a large
+            # number of connections, which is not suitable for testing.
+            set size 400
+            while {$size <= 8000} {
+                lappend channel_sizes $size
+                dict set channel_name_dict $size [string repeat x $size]
+                incr size 50
+            }
+
+            # Populate memory with interleaving key-pubsub pattern of variable sizes
+            set n 100
+            set rd [redis_deferring_client]
+            set rds_pubsub []
+            for {set j 0} {$j < $n} {incr j} {
+                set channels {}
+                # Fill them with keys of various sizes
+                foreach size $channel_sizes {
+                    set val "[dict get $channel_name_dict $size][format "%06d" $j]"
+                    $rd set k${j}_$size $val
+                    lappend channels $val
+                }
+
+                # Fill in the gaps behind the keys.
+                set rd_pubsub [redis_deferring_client]
+                $rd_pubsub subscribe {*}$channels
+                $rd_pubsub read
+                lappend rds_pubsub $rd_pubsub
+            }
+            for {set j 0} {$j < [expr $n * [llength $channel_sizes]]} {incr j} {
+                $rd read ; # Discard set replies
+            }
+            after 120 ;# serverCron only updates the info once in 100ms
+            if {$::verbose} {
+                puts "used [s allocator_allocated]"
+                puts "rss [s allocator_active]"
+                puts "frag [s allocator_frag_ratio]"
+                puts "frag_bytes [s allocator_frag_bytes]"
+            }
+            assert_lessthan [s allocator_frag_ratio] 1.05
+
+            # Delete all the keys to create fragmentation
+            for {set j 0} {$j < $n} {incr j} { 
+                foreach size $channel_sizes {
+                    $rd del k${j}_$size
+                }
+            }
+            for {set j 0} {$j < [expr $n * [llength $channel_sizes]]} {incr j} { $rd read } ; # Discard del replies
+            $rd close
+            after 120 ;# serverCron only updates the info once in 100ms
+            if {$::verbose} {
+                puts "used [s allocator_allocated]"
+                puts "rss [s allocator_active]"
+                puts "frag [s allocator_frag_ratio]"
+                puts "frag_bytes [s allocator_frag_bytes]"
+            }
+            assert_morethan [s allocator_frag_ratio] 1.15
```

Comment:
please note that i used `1.15` instead `1.4` as defragging scripts.
because scripts defragment uses a single size key, but in the pubsub test we can't set the key and then load the script as it were, which would require a lot of connections and would not be suitable for testing.
so I use a client that subscribes to various channels with various key sizes after inserting string with various sizes, to avoid a lot of connections.

Reply:
i think we can do better.
we can generate lots of different names of the same size instead of creating a ton of different sizes.
e.g. we can aim for the 640 bytes bin.
take a prefix of 600 static characters, and then add an `%d` suffix.

Reply:
I've actually tried, in theory, if we use the class 640, allocating 50 keys with 640 bytes, then subscribing to 50 channels with 640 bytes, and then deleting those 50 keys should yield about 50% fragmentation, not sure why that can't be expected.
Am I missing something here?

Reply:
i didn't understand your last message at all.
we need to keep in mind that the fragmentation ration is relative to the consumed memory, so even if we generate high fragmentation on the 640 bytes bin, but we also generate high memory consumption (e.g. 32kb allocations), the fragmentation we generated will not show high percentage.
i suppose that's why we need to avoid using many clients (each has it's buffers, and a big client struct).
but i won't understand why we can't generate thousands of subscriptions with the same client and a very similar channel name (length wise).

Reply:
for example, when a client inserts 10,000 keys with 1024 bytes, then another client subscribes to 10,000 channels with different names of 1024 bytes, and then deletes those 10,000 keys, the fragmentation rate is actually close to zero.
because the keys and channel names are always on separate pages, that is, the reg of keys and the channel names are never in the same page, when all the keys are deleted, the pages using by the keys will be reclaimed.

Reply:
that depends on the order in which you do the subscribe and key creation. if you interlace them (create a key, then subscribe to a channel, then loop), they'll share pages.
of am i missing some point?

Reply:
silly me, i'm missing that we can subscribe again in pubsub state.

Reply:
done with [`cf8694e` (#13058)](https://github.com/redis/redis/pull/13058/commits/cf8694e571da5a0a814e88e38da36282f0ce9e12)